finagle-http/src/main/scala/com/twitter/finagle/http/Request.scala

```diff
@@ -287,10 +287,36 @@ object Request {
    */
   def apply(version: Version, method: Method, uri: String): Request = {
     val reqIn = new DefaultHttpRequest(from(version), from(method), uri)
-    new Request {
-      val httpRequest: HttpRequest = reqIn
-      lazy val remoteSocketAddress: InetSocketAddress = new InetSocketAddress(0)
+    buildRequest(
+      request = reqIn,
+      socketAddress = new InetSocketAddress(0)
+    )
+  }
+
+  private def buildRequest(
+    request: HttpRequest,
+    socketAddress: InetSocketAddress,
+    reader: Option[Reader] = None): Request = {
+
+    val req = reader match {
+      case Some(readerIn: Reader) =>
+        new  Request {
+          override val reader: Reader = readerIn
+          val httpRequest: HttpRequest = request
+          lazy val remoteSocketAddress: InetSocketAddress = socketAddress
+        }
+      case None =>
+        new  Request {
+          val httpRequest: HttpRequest = request
+          lazy val remoteSocketAddress: InetSocketAddress = socketAddress
+        }
+    }
+
+    if (req.version == Version.Http11 && !req.headers.contains(HttpHeaders.Names.HOST)) {
+      req.headers.set(HttpHeaders.Names.HOST, "")
```

Comment:
I already added link to RFC in my commit. Here a quote from it:
```
 A client MUST send a Host header field in all HTTP/1.1 request
   messages.  If the target URI includes an authority component, then a
   client MUST send a field-value for Host that is identical to that
   authority component, excluding any userinfo subcomponent and its "@"
   delimiter (Section 2.7.1).  If the authority component is missing or
   undefined for the target URI, then a client MUST send a Host header
   field with an empty field-value.
```
I think an empty Host header during initialization is ok.
 Other way is to parse a Host from URI like you already do in `RequestBuilder`. And then we must remove this logic from `RequestBuilder` to avoid duplications.  But I did not find a way to do it neatly and without changes in public interfaces.